src/Swarm/Game/ScenarioInfo.hs

```diff
@@ -153,22 +160,37 @@ scenarioBestTime :: Lens' ScenarioInfo ScenarioStatus
 -- | The best status of the scenario, measured in game ticks.
 scenarioBestTicks :: Lens' ScenarioInfo ScenarioStatus
 
+-- | The best code size of the scenario, measured both in character count and AST size.
+scenarioBestCodeSize :: Lens' ScenarioInfo (Maybe ScenarioCodeMetrics)
+
 -- | Update the current @ScenarioInfo@ record when quitting a game.
 --
 -- Note that when comparing "best" times, shorter is not always better!
 -- As long as the scenario is not completed (e.g. some do not have win condition)
 -- we consider having fun _longer_ to be better.
-updateScenarioInfoOnQuit :: ZonedTime -> Integer -> Bool -> ScenarioInfo -> ScenarioInfo
-updateScenarioInfoOnQuit z ticks completed (ScenarioInfo p s bTime bTicks) = case s of
-  InProgress start _ _ ->
-    let el = (diffUTCTime `on` zonedTimeToUTC) z start
-        cur = (if completed then Complete else InProgress) start el ticks
-        best f b = case b of
-          Complete {} | not completed || f b <= f cur -> b -- keep faster completed
-          InProgress {} | not completed && f b > f cur -> b -- keep longer progress (fun!)
-          _ -> cur -- otherwise update with current
-     in ScenarioInfo p cur (best _scenarioElapsed bTime) (best _scenarioElapsedTicks bTicks)
-  _ -> error "Logical error: trying to quit scenario which is not in progress!"
+updateScenarioInfoOnQuit ::
+  CodeSizeDeterminators ->
+  ZonedTime ->
+  Integer ->
+  Bool ->
+  ScenarioInfo ->
+  ScenarioInfo
+updateScenarioInfoOnQuit
+  csd
+  z
+  ticks
+  completed
+  (ScenarioInfo p s bTime bTicks _prevCodeSize) = case s of
```

Comment:
TODO: need to take `_prevCodeSize` into account and use the smaller of the two.

Reply:
Sounds like a `Semigroup` on `ScenarioCodeMetrics`.